tracing-subscriber/src/filter/env/directive.rs

```diff
@@ -120,99 +118,120 @@ impl Directive {
     }
 
     pub(super) fn parse(from: &str, regex: bool) -> Result<Self, ParseError> {
-        static DIRECTIVE_RE: Lazy<Regex> = Lazy::new(|| {
-            Regex::new(
-                r"(?x)
-            ^(?P<global_level>(?i:trace|debug|info|warn|error|off|[0-5]))$ |
-                #                 ^^^.
-                #                     `note: we match log level names case-insensitively
-            ^
-            (?: # target name or span name
-                (?P<target>[\w:-]+)|(?P<span>\[[^\]]*\])
-            ){1,2}
-            (?: # level or nothing
-                =(?P<level>(?i:trace|debug|info|warn|error|off|[0-5]))?
-                    #          ^^^.
-                    #              `note: we match log level names case-insensitively
-            )?
-            $
-            ",
-            )
-            .unwrap()
-        });
-        static SPAN_PART_RE: Lazy<Regex> =
-            Lazy::new(|| Regex::new(r"(?P<name>[^\]\{]+)?(?:\{(?P<fields>[^\}]*)\})?").unwrap());
-        static FIELD_FILTER_RE: Lazy<Regex> =
-            // TODO(eliza): this doesn't _currently_ handle value matchers that include comma
-            // characters. We should fix that.
-            Lazy::new(|| {
-                Regex::new(
-                    r"(?x)
-                (
-                    # field name
-                    [[:word:]][[[:word:]]\.]*
-                    # value part (optional)
-                    (?:=[^,]+)?
-                )
-                # trailing comma or EOS
-                (?:,\s?|$)
-            ",
-                )
-                .unwrap()
-            });
-
-        let caps = DIRECTIVE_RE.captures(from).ok_or_else(ParseError::new)?;
+        let mut cur = Self {
+            level: LevelFilter::TRACE,
+            target: None,
+            in_span: None,
+            fields: Vec::new(),
+        };
+
+        #[derive(Debug)]
+        enum ParseState {
+            Start,
+            LevelOrTarget { start: usize },
+            Span { span_start: usize },
+            Field { field_start: usize },
+            Fields,
+            Target,
+            Level { level_start: usize },
+            Complete,
+        }
 
-        if let Some(level) = caps
-            .name("global_level")
-            .and_then(|s| s.as_str().parse().ok())
-        {
-            return Ok(Directive {
-                level,
-                ..Default::default()
-            });
+        use ParseState::*;
+        let mut state = Start;
+        for (i, c) in from.trim().char_indices() {
+            state = match (state, c) {
+                (Start, '[') => Span { span_start: i + 1 },
+                (Start, c) if c.is_ascii_punctuation() => return Err(ParseError::new()),
```

Comment:
I added this arm to pass the new test cases from #3262. This feels like a bit of a hack because while these cases used ASCII punctuation to make the directive invalid, I suppose other characters could do so too. At the same time, I think we want to allow non-ASCII identifier characters here so we can't be too strict either, and it'd be nice to avoid ballooning complexity here. Maybe it would be better to use (non-ASCII) `!is_alphanumeric()` here?

Reply:
Changed my own mind, see https://github.com/tokio-rs/tracing/pull/3243#discussion_r2069054929 instead.